controllers/argocd/keycloak.go

```diff
@@ -1222,6 +1296,20 @@ func (r *ReconcileArgoCD) reconcileKeycloakForOpenShift(cr *argoprojv1a1.ArgoCD)
 	if err != nil {
```

Comment:
You mentioned that it was not possible to write tests because these new features depeend on UI. Is it possible to write fake reconciler test to validate these reconcilation logic?

Reply:
@wtam2018 Let me try to explain you in more detail way on **why I think** the features in the PR is **difficult(if not imposible) to test without UI.**

**Using Testing or golang test framework:**
Steps to acheive :
 - Create a Fake OpenShift client
 - Create a Fake Keycloak client
 - make an API call to the Fake Keycloak client to create the OpenShift IdP(initializing fake OpenShift client)
 - Setup Identity Provider on you OpenShift cluster to create users assign them to groups.
 - The most difficult part, Steps from the UI -> Login to Argo CD using the link `Login via Keycloak` -> `Login with OpenShift` -> Provide OpenShift Credentials -> Submit User information like Email -> Verify the logged in User details from Argo CD console.
 
 Please let me know if you can think of someway to achieve this without testing it from UI, I can think of using `argocd` cli binary but that can only help us to some extent.
 
 @jannfis  any help here ?

Reply:
I think we can test for any reconciliation logic using kuttl, but for testing the actual login procedure, we'd need some UI testing framework in place (which we don't have yet).

Reply:
We should at least add unit tests. This is a sizable PR with zero test coverage.

Reply:
@wtam2018 I looked into the PR one more time. Apart from `isProxyCluster()` all the other code in this PR is
1. Image upgrade
2. New Type Declarations
3. Doc updates
4. bundle & manifest changes.

I will try to create a fakeOpenShift client and add unittests to test the `isProxyCluster()`. I will see if I can add more unit tests to overall keycloak functionality. Thanks :)

Argo CD Operator generally does not have kuttl tests(e2e tests) for keycloak/RHSSO. we have some in GitLab regression suite. we should some e2e tests in this operator.

Reply:
@wtam2018 
I have added  e2e tests for rhsso 
------------

Download kuttl on your laptop or server
Run operator locally or through the bundle.
Run the below command to execute rhsso e2e tests
kubectl kuttl test --config kuttl-test-rhsso.yaml